### PR TITLE
Updating Task Model and implementing docker flags --pid and --ipc

### PR DESCRIPTION
**Pull request by linkar-ec2**
### Summary

Updates the Agent Task model to include PidMode and IpcMode for sharing PID namespace and IPC resource namespace, respectively. 
Adding unit tests for Pid and Ipc resource namespace sharing

### Implementation details

Agent Task struct is updated to support the PidMode and IpcMode that is in the Task struct received in the payload message from ACS. 
In PostUnmarshalTask(), the task is used to determine if a Pause container for namespace sharing needs to be provisioned. If so, it is added as an internal container to the task. 
In CreateContainer(), the DockerHostConfig() method determines if the current container's PidMode and IpcOption needs to be changed. 

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 
Unit Tests cover the added Pause container provisioning methods and DockerHostConfig updates
### Description for the changelog

Adding ECS Agent support for the --pid and --ipc Docker flags
### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by samuelkarp** [comment]
Is there any particular reason that you're not sharing the pause container between awsvpc mode and these new options?  They serve the same purpose: to provide root namespaces that can be shared.

**Comment by linkar-ec2** [comment]
The purpose of the ENI Pause Container was to attach the individual containers within a task to an ENI. We did this by creating a common network namespace and then attaching the Pause container to the ENI.

Reusing this Pause container would tightly couple this ENI implementation with the namespace sharing for PID and IPC resources. With the resources consumed by the Pause container considered to be negligible, creating another container would allow single responsibility of the internal containers we create.  A concrete example of this would be the case when an ENI has custom DNS IPs and search list associated with it, in which case we will configure the HostConfig for the ENI Pause, but not for the PID/IPC namespace Pause container.

As regards to the possible question “why not create 2 Pause containers for the PID and IPC resources for this same reason of single responsibility?” this Namespace Pause container is simply provided as a parameter to the other containers within a task, for either PidMode or IpcMode, or both. There is no other HostConfig changes done to the Pause container itself (a difference from the ENI Pause).

Hope this sheds some light. Please don’t hesitate to ask questions or comments.

Best,

Kar

**Comment by samuelkarp** [comment]
That's not exactly accurate.  The purpose of the existing pause container is to create a network namespace that can be shared among the containers in a task.  We don't attach the pause container to the ENI, but rather move the ENI into the network namespace that was created as a side-effect of running the pause container.  PID and IPC namespaces are similar in that they're just namespaces of the container.

File: agent/api/task/task.go [hunk 1]
```diff
@@ -536,11 +551,37 @@ func (task *Task) addNetworkResourceProvisioningDependency(cfg *config.Config) {
...
+	if task.getIpcMode() == "" && task.getPidMode() == "" {
+		return
+	}
+
+	// Create a Pause container if either IPC resource or PID namespace needs to be shared
+	if task.getIpcMode() == "task" || task.getPidMode() == "task" {
+		namespacePauseContainer := apicontainer.NewContainerWithSteadyState(apicontainerstatus.ContainerResourcesProvisioned)
+		namespacePauseContainer.TransitionDependenciesMap = make(map[apicontainerstatus.ContainerStatus]apicontainer.TransitionDependencySet)
+		namespacePauseContainer.Name = NamespacePauseContainerName
+		namespacePauseContainer.Image = fmt.Sprintf("%s:%s", cfg.PauseContainerImageName, cfg.PauseContainerTag)
```
Comment (haikuoliu) [inline agent/api/task/task.go]: To avoid `PauseContainerImageName` being overridden by Fargate, one option is you can use `DefaultPauseContainerImageName` here. And you will also need to make change [here](https://github.com/aws/amazon-ecs-agent/blob/e8b5229a9ea8b748ebad4db4c098ccc8df2dcd5f/agent/app/agent_unix.go#L84) and `pause_linux.go` to make Agent always load *default* pause container tarball

**Comment by linkar-ec2** [comment]
[After syncing with Sam and Derek] 
For closure, it is better to use separate Pause containers for the ENI/Network sharing and the PID/IPC resources namespace sharing. The image for the Pause container for the ENI attachment is specific to the requirements of the ENI. The image for the PID/IPC sharing Pause will be the default image.

File: agent/api/task/task.go [hunk 2]
```diff
@@ -831,21 +872,30 @@ func (task *Task) dockerHostConfig(container *apicontainer.Container, dockerCont
...
+			if hosts := task.generateENIExtraHosts(); hosts != nil {
+				hostConfig.ExtraHosts = append(hostConfig.ExtraHosts, hosts...)
+			}
 
-		// apply ExtraHosts to HostConfig for pause container
-		if hosts := task.generateENIExtraHosts(); hosts != nil {
-			hostConfig.ExtraHosts = append(hostConfig.ExtraHosts, hosts...)
+			// Override the DNS settings for the pause container if ENI has custom
+			// DNS settings
+			return task.overrideDNS(hostConfig), nil
```
[the remark this answers is not in the document]
Comment (linkar-ec2) [reply agent/api/task/task.go]: This return specifically applies to the CNIPauseContainer. Because we are using different Pause containers, we do not want to apply the Pid or Ipc settings to this specific Docker container. Therefore, our expected call flow would be to turn here.

File: agent/api/container/container.go [hunk 3]
```diff
@@ -401,8 +401,8 @@ func (c *Container) GetNextKnownStateProgression() apicontainerstatus.ContainerS
 	return c.GetKnownStatus() + 1
 }
 
-// IsInternal returns true if the container type is either `ContainerEmptyHostVolume`
-// or `ContainerCNIPause`. It returns false otherwise
+// IsInternal returns true if the container type is `ContainerEmptyHostVolume`,
+// `ContainerCNIPause`, or `ContainerNamespacePause`. It returns false otherwise
```
[the remark this answers is not in the document]
Comment (linkar-ec2) [reply agent/api/container/container.go]: Because the ContainerNamespacePause is used for both IPC resource namespace sharing and PID namespace sharing, it was named with 'Namespace' rather than 'Process'

File: agent/api/task/task.go [hunk 4]
```diff
@@ -953,6 +1004,96 @@ func (task *Task) generateENIExtraHosts() []string {
...
+		return false, ""
+	}
+
+	// no need to override the PID mode. This means the container's PID namespace
+	// will be private
+	if task.getPIDMode() == "" || strings.EqualFold(task.getPIDMode(), "none") {
+		return false, ""
+	}
+
+	if strings.EqualFold(task.getPIDMode(), "host") {
```
[the remark this answers is not in the document]
Comment (linkar-ec2) [reply agent/api/task/task.go]: @yunhee-l Functional tests need the Frontend changes to be complete before we can register task definitions. Integ tests are currently being made and will probably be in another PR

File: agent/api/task/task.go [hunk 5]
```diff
@@ -1016,28 +1017,21 @@ func (task *Task) shouldOverridePidMode(container *apicontainer.Container, docke
...
-		for _, cont := range task.Containers {
-			if cont.Type == apicontainer.ContainerNamespacePause {
-				pidPauseContName = cont.Name
-				break
-			}
-		}
-		if pidPauseContName == "" {
-			seelog.Critical("Namespace Pause Container required, but not found in the task: %s", task.String())
-			return false, "" // cannot override the Pid without the Pause container
+	} else if strings.EqualFold(task.getPIDMode(), "task") {
```
Comment (haikuoliu) [inline agent/api/task/task.go]: Any particular reason using `strings.EqualFold` instead of `==`? I think `==` should be sufficient here as we already know the result of `task.getPIDMode()`, i.e. it's `"task"` instead of `"Task"` or `"tAsk"`, this is more like a contract between Agent and ACS.

File: agent/app/agent_unix.go [hunk 6]
```diff
@@ -81,17 +81,15 @@ func (agent *ecsAgent) initializeTaskENIDependencies(state dockerstate.TaskEngin
 		return err, true
 	}
 
-	if agent.cfg.ShouldLoadPauseContainerTarball() {
```
Comment (haikuoliu) [inline agent/app/agent_unix.go]: Since we will load the default pause container anyway, `ShouldLoadPauseContainerTarball` loses its meaning, the only consumer of this function is `platformString`, can you please name it to something that matches its meaning? Or you can just delete this function and move the logic into `platformString`

File: agent/api/container/container.go [hunk 7]
```diff
@@ -401,8 +401,8 @@ func (c *Container) GetNextKnownStateProgression() apicontainerstatus.ContainerS
 	return c.GetKnownStatus() + 1
 }
 
-// IsInternal returns true if the container type is either `ContainerEmptyHostVolume`
-// or `ContainerCNIPause`. It returns false otherwise
+// IsInternal returns true if the container type is `ContainerEmptyHostVolume`,
```
Comment (sharanyad) [inline agent/api/container/container.go]: We don't use the `ContainerEmptyHostVolume` anymore, it can be removed from the comment.

File: agent/api/task/task.go [hunk 8]
```diff
@@ -536,11 +561,33 @@ func (task *Task) addNetworkResourceProvisioningDependency(cfg *config.Config) {
...
 		pauseContainer.BuildContainerDependency(container.Name, apicontainerstatus.ContainerStopped, apicontainerstatus.ContainerStopped)
 	}
 }
 
+func (task *Task) addNamespaceSharingProvisioningDependency(cfg *config.Config) {
+	// Pause container does not need to be created if no namespace sharing will be done at task level
+	if task.getIPCMode() != ipcModeTask && task.getPIDMode() != pidModeTask {
+		return
+	}
+	namespacePauseContainer := apicontainer.NewContainerWithSteadyState(apicontainerstatus.ContainerResourcesProvisioned)
```
Comment (sharanyad) [inline agent/api/task/task.go]: What is the definition of `ContainerResourcesProvisioned` state for namespacePauseContainer?

File: agent/api/task/task.go [hunk 9]
```diff
@@ -953,6 +1014,91 @@ func (task *Task) generateENIExtraHosts() []string {
...
+// false otherwise
+func (task *Task) shouldOverridePIDMode(container *apicontainer.Container, dockerContainerMap map[string]*apicontainer.DockerContainer) (bool, string) {
+	// If the container is an internal container (ContainerEmptyHostVolume,
+	// ContainerCNIPause, or ContainerNamespacePause), then PID namespace for
+	// the container itself should be private (default Docker option)
+	if container.IsInternal() {
+		return false, ""
+	}
+
+	if task.getPIDMode() == pidModeHost {
```
Comment (sharanyad) [inline agent/api/task/task.go]: These can be switch statements instead of if, else. That would be easier to read.

File: agent/api/task/task.go [hunk 10]
```diff
@@ -953,6 +1014,91 @@ func (task *Task) generateENIExtraHosts() []string {
...
+		return true, pidModeHost
+	} else if task.getPIDMode() == pidModeTask {
+		pauseContName, ok := task.findPauseContainerName(apicontainer.ContainerNamespacePause)
+		if !ok {
+			seelog.Critical("Namespace Pause container not found in the task: %s", task.String())
+			return false, ""
+		}
+		pauseToGiveDocker, ok := dockerContainerMap[pauseContName]
```
Comment (sharanyad) [inline agent/api/task/task.go]: what does `pauseToGiveDocker` mean?

File: agent/api/task/task.go [hunk 11]
```diff
@@ -953,6 +1014,91 @@ func (task *Task) generateENIExtraHosts() []string {
...
+		if !ok || pauseToGiveDocker == nil {
+			// Docker container shouldn't be nill or not exist if the Container definition within task exists; implies code-bug
```
Comment (sharanyad) [inline agent/api/task/task.go]: nit: `nil`

File: agent/api/task/task.go [hunk 12]
```diff
@@ -953,6 +1014,91 @@ func (task *Task) generateENIExtraHosts() []string {
...
+			seelog.Criticalf("Namespace Pause container required, but not found in container map for container: [%s] in task: %s",
```
Comment (sharanyad) [inline agent/api/task/task.go]: This is a failure case, which should be handled in a way that the task is not run. If pidMode is task and pause container is not found, we should stop the task imo. same for ipcMode.

File: agent/api/task/task.go (hunk 10 above)
Comment (linkar-ec2) [reply agent/api/task/task.go]: This is the Docker container name that is contained in the map. We will 'give' this container to the Docker api for the --pid or --ipc flag

File: agent/api/task/task.go (hunk 9 above)
Comment (linkar-ec2) [reply agent/api/task/task.go]: The first if block tests if the container is internal. This is unrelated to the Task's pidMode, whose logic is an if and else-if block. For separation of logic, I think this if-else works better than the switch. What do you think?

File: agent/api/task/task.go (hunk 10 above)
Comment (sharanyad) [reply agent/api/task/task.go]: Got it. it's a little confusing. I would prefer just calling it pauseContainerID or something similar.

Comment (linkar-ec2) [reply agent/api/task/task.go]: Changed!

File: agent/api/task/task.go (hunk 8 above)
Comment (linkar-ec2) [reply agent/api/task/task.go]: ContainerResourcesProvisioned seems to be the norm for internal containers. Even though this state checks to see if ENI resources need to be added (for our namespacePauseContainer, nothing additional will be done). Because we don't need anything in this method anyways, I set the Dependency to use Running. I don't think there should be difference in runtime though.

File: agent/api/task/task.go [hunk 13]
```diff
@@ -583,7 +583,7 @@ func (task *Task) addNamespaceSharingProvisioningDependency(cfg *config.Config)
 		if container.IsInternal() {
 			continue
 		}
-		container.BuildContainerDependency(NamespacePauseContainerName, apicontainerstatus.ContainerResourcesProvisioned, apicontainerstatus.ContainerPulled)
+		container.BuildContainerDependency(NamespacePauseContainerName, apicontainerstatus.ContainerRunning, apicontainerstatus.ContainerPulled)
```
Comment (haikuoliu) [inline agent/api/task/task.go]: You will also need to change 
```
namespacePauseContainer := apicontainer.NewContainerWithSteadyState(apicontainerstatus.ContainerResourcesProvisioned)
``` 
to 
```
namespacePauseContainer := apicontainer.NewContainerWithSteadyState(apicontainerstatus.ContainerRunning)
```
Otherwise the Agent will call `provisionContainerResources` to setup CNI for namespace pause container.

File: agent/api/task/task.go [hunk 14]
```diff
@@ -76,6 +76,16 @@ const (
 	// awslogsCredsEndpointOpt is the awslogs option that is used to pass in an
 	// http endpoint for authentication
 	awslogsCredsEndpointOpt = "awslogs-credentials-endpoint"
+
+	// ipcModeNone specifies the string used to define the `none` docker IPC resource namespace sharing mode
```
Comment (haikuoliu) [inline agent/api/task/task.go]: Any reason that `ipcModeNone` is not grouped with other pid/ipc constants?

File: agent/api/task/task.go (hunk 13 above)
Comment (linkar-ec2) [reply agent/api/task/task.go]: Good catch. Updated.

File: agent/api/task/task.go (hunk 14 above)
Comment (linkar-ec2) [reply agent/api/task/task.go]: Just to distinguish that None is an option for IPC but thinking again, it makes more sense to just group it and leave out the comment.

File: agent/api/task/task_test.go [hunk 15]
```diff
@@ -635,6 +635,183 @@ func TestPostUnmarshalTaskWithLocalVolumes(t *testing.T) {
 
 }
 
+// Slice of structs for Table Driven testing for sharing PID and IPC resources
+var namespaceTests = []struct {
```
Comment (sharanyad) [inline agent/api/task/task_test.go]: Please add tests for IPCMode None and Shareable as well.

File: agent/api/task/task_test.go [hunk 16]
```diff
@@ -635,6 +635,183 @@ func TestPostUnmarshalTaskWithLocalVolumes(t *testing.T) {
...
+					case "host":
+						assert.Equal(t, "host", dockHostCfg.PidMode)
+					}
+				}
+			}
+		}
+	}
+}
+
+func TestAddNamespaceSharingProvisioningDependency(t *testing.T) {
```
Comment (sharanyad) [inline agent/api/task/task_test.go]: I think this test is missing assertion of the fields of `TransitionDependenciesMap`, which is populated by the method `addNamespaceSharingProvisioningDependency`

File: agent/api/task/task_test.go (hunk 15 above)
Comment (linkar-ec2) [reply agent/api/task/task_test.go]: The customer will not be allowed to provide Shareable as an option. Shareable is internally set for the Pause container and is tested within a unit test.

File: agent/api/task/task_test.go (hunk 16 above)
Comment (linkar-ec2) [reply agent/api/task/task_test.go]: Added!

**Comment by linkar-ec2** [comment]
Code Coverage: 
Task: 70.1%
Container: 57.3%

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,4 +1,9 @@
 # Changelog
+## 1.20.4-dev
```
Comment (sharanyad) [inline CHANGELOG.md]: Looks like this file should not be there in this PR.

Comment (linkar-ec2) [reply CHANGELOG.md]: I rebased and squashed all commits. Awaiting merge.

File: agent/api/task/task.go [hunk 18]
```diff
@@ -953,6 +1006,95 @@ func (task *Task) generateENIExtraHosts() []string {
...
+		pauseCont, ok := task.ContainerByName(NamespacePauseContainerName)
+		if !ok {
+			seelog.Criticalf("Namespace Pause container not found in the task: %s; Setting Task's Desired Status to Stopped", task.Arn)
+			task.SetDesiredStatus(apitaskstatus.TaskStopped)
+			return false, ""
+		}
+		pauseDockerID, ok := dockerContainerMap[pauseCont.Name]
+		if !ok || pauseDockerID == nil {
+			// Docker container shouldn't be nil or not exist if the Container definition within task exists; implies code-bug
+			seelog.Criticalf("Namespace Pause container not found in the task: %s; Setting Task's Desired Status to Stopped", task.Arn)
```
Comment (sharanyad) [inline agent/api/task/task.go]: nit: this error message is same as the previous condition. 
Please change it to something like not found in container map, so its easier while debugging.
